aiozk/connection.py

```diff
@@ -258,9 +258,14 @@ def drain_all_pending(self):
     async def close(self, timeout):
         if self.closing:
             return
+        self.closing = True
+        if self.read_loop_task:
+            try:
+                await asyncio.wait_for(self.read_loop_task, 0.1, loop=self.loop)
```

Comment:
@cybergrind what do you think about it?

I am using `asyncio.wait_for` here because `xid, zxid, response = await self.read_response()` within `Connection.read_loop` is blocking and cannot be simply cancelled without timeout.

Reply:
Is there something special about `await self.read_response()`, because you cannot cancel tasks in such way only if task caught `CancelledError` silently, but I cannot spot this in `asyncio.streams`. Basically `StreamReader` in `.read` method just wait for _waiter, and it is a Future, so it cannot prevent CancelledError

Maybe we need test case?

Also, I'm not sure why `while True` is better than `while not self.closing`.

Reply:
@cybergrind `StreamReader._waiter` cannot be woken up as we don't get any data or EOF so it blocks.

Concerning `while True` I think that using `while not self.closing` is redundant if task is being cancelled explicitly: in case of `task.cancel()` "while loop" breaks and `self.closing` even not evaluated.

Reply:
`StreamReader._waiter` is a `Future` and I cannot find any mentions that we can have a situation when we cannot cancel a `Future`.

> StreamReader._waiter cannot be woken up as we don't get any data or EOF so it blocks.

It doesn't make any sense for me because I cannot find any blocking code. Am I missing something?

Reply:
@cybergrind if `StreamReader._waiter` is cancelled then `StreamReader._wakeup_waiter` cannot set result for it therefore `StreamReader._wait_for_data` stucks awaiting cancelled `StreamReader._waiter` future.

Reply:
I'm not sure that situation is caused by such flow - CancelledErrors are propagated through all connected tasks.

At least if we just add `await asyncio.sleep(0.0001)`at the end of close function - it will work without `wait_for`. So it seems that task.cancel doesn't immediately cancels tasks but schedules cancellation to next loop and we just need to give to event loop this extra loop before call event_loop.stop()

Reply:
```python
async def stop_handler():
    await zk.close()
    loop.call_soon(loop.stop)
```
This also can fix the suggested test.

Reply:
Yes, I noticed this behavior. And we even don't need to cancel `read_loop_task` before awaiting some `asyncio.sleep` within `Connection.close()`.

`loop.call_soon(loop.stop)` can fix the problem but it is external fix - the problem will remain internally within `aiozk`. I think the "extra loop" should be awaited within `Connection.close()`.